File: backend/app/api/routes/templates.py
"""
Template API Routes

API endpoints for campaign template management.
"""
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session
from sqlalchemy import select, func

from app.api.dependencies.auth import get_current_master_admin_user
from app.db.session import get_db
from app.models.user import User
from app.models.master_admin import CampaignTemplate
from app.models.organization import Organization
from app.schemas.master_admin import (
    CampaignTemplateCreate,
    CampaignTemplateUpdate,
    CampaignTemplateResponse,
    CampaignTemplateListResponse,
    TemplatePreviewRequest,
    TemplatePreviewResponse,
)
from app.services import template_service

router = APIRouter(prefix="/master-admin/templates", tags=["templates"])


@router.post("", response_model=CampaignTemplateResponse, status_code=status.HTTP_201_CREATED)
def create_template(
    template_data: CampaignTemplateCreate,
    current_user: User = Depends(get_current_master_admin_user),
    db: Session = Depends(get_db),
):
    """
    Create a new campaign template.
    
    Args:
        template_data: Template creation data
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Created template
    """
    org_result = db.execute(
        select(Organization).where(Organization.id == current_user.organization_id)
    )
    organization = org_result.scalar_one_or_none()
    
    if not organization:
        raise HTTPException(status_code=404, detail="Organization not found")
    
    template_dict = template_data.model_dump()
    template = template_service.create_template(
        template_dict,
        str(organization.id),
        str(current_user.id),
        db
    )
    return template


@router.get("", response_model=CampaignTemplateListResponse)
def list_templates(
    type_filter: Optional[str] = Query(None, alias="type"),
    is_default: Optional[bool] = Query(None),
    current_user: User = Depends(get_current_master_admin_user),
    db: Session = Depends(get_db),
):
    """
    List templates for the current user's organization.
    
    Args:
        type_filter: Filter by template type
        is_default: Filter by is_default flag
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        List of templates
    """
    org_result = db.execute(
        select(Organization).where(Organization.id == current_user.organization_id)
    )
    organization = org_result.scalar_one_or_none()
    
    if not organization:
        raise HTTPException(status_code=404, detail="Organization not found")
    
    templates = template_service.list_templates(
        str(organization.id),
        db,
        template_type=type_filter,
        is_default=is_default
    )
    
    return {
        "items": templates,
        "total": len(templates),
    }


@router.get("/{template_id}", response_model=CampaignTemplateResponse)
def get_template(
    template_id: int,
    current_user: User = Depends(get_current_master_admin_user),
    db: Session = Depends(get_db),
):
    """
    Get a specific template.
    
    Args:
        template_id: Template ID
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Template details
        
    Raises:
        HTTPException: If template not found
    """
    org_result = db.execute(
        select(Organization).where(Organization.id == current_user.organization_id)
    )
    organization = org_result.scalar_one_or_none()
    
    if not organization:
        raise HTTPException(status_code=404, detail="Organization not found")
    
    template = template_service.get_template(template_id, str(organization.id), db)
    
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    
    return template


@router.put("/{template_id}", response_model=CampaignTemplateResponse)
def update_template(
    template_id: int,
    template_update: CampaignTemplateUpdate,
    current_user: User = Depends(get_current_master_admin_user),
    db: Session = Depends(get_db),
):
    """
    Update a template.
    
    Args:
        template_id: Template ID
        template_update: Template update data
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Updated template
        
    Raises:
        HTTPException: If template not found
    """
    org_result = db.execute(
        select(Organization).where(Organization.id == current_user.organization_id)
    )
    organization = org_result.scalar_one_or_none()
    
    if not organization:
        raise HTTPException(status_code=404, detail="Organization not found")
    
    template = template_service.get_template(template_id, str(organization.id), db)
    
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    
    update_data = template_update.model_dump(exclude_unset=True)
    template = template_service.update_template(template, update_data, db)
    
    return template


@router.delete("/{template_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_template(
    template_id: int,
    current_user: User = Depends(get_current_master_admin_user),
    db: Session = Depends(get_db),
):
    """
    Delete a template.
    
    Args:
        template_id: Template ID
        current_user: Current authenticated user
        db: Database session
        
    Raises:
        HTTPException: If template not found
    """
    org_result = db.execute(
        select(Organization).where(Organization.id == current_user.organization_id)
    )
    organization = org_result.scalar_one_or_none()
    
    if not organization:
        raise HTTPException(status_code=404, detail="Organization not found")
    
    template = template_service.get_template(template_id, str(organization.id), db)
    
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    
    template_service.delete_template(template, db)


@router.post("/{template_id}/preview", response_model=TemplatePreviewResponse)
def render_template_preview(
    template_id: int,
    preview_request: TemplatePreviewRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Preview a template with sample contact data.
    
    Args:
        template_id: Template ID
        preview_request: Request with contact_data for variable substitution
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        Rendered template (subject and content)
        
    Raises:
        HTTPException: If template not found or variables missing
    """
    org_result = db.execute(
        select(Organization).where(Organization.id == current_user.organization_id)
    )
    organization = org_result.scalar_one_or_none()
    
    if not organization:
        raise HTTPException(status_code=404, detail="Organization not found")
    
    template = template_service.get_template(template_id, str(organization.id), db)
    
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    
    try:
        rendered = template_service.render_template(
            template_id,
            preview_request.contact_data,
            db
        )
        return rendered
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

File: frontend/src/services/api/templates.ts
/**
 * Template API Client
 * 
 * TypeScript client for campaign template API endpoints.
 */
import { apiClient } from './client'

// Types
export interface CampaignTemplate {
  id: number
  organization_id: string
  name: string
  subject?: string
  content: string
  type: 'email' | 'voice' | 'linkedin' | 'multi_channel'
  variables?: string[]
  is_default: boolean
  created_by?: string
  created_at: string
  updated_at: string
}

export interface CampaignTemplateCreate {
  name: string
  subject?: string
  content: string
  type: 'email' | 'voice' | 'linkedin' | 'multi_channel'
  variables?: string[]
  is_default?: boolean
}

export interface CampaignTemplateUpdate {
  name?: string
  subject?: string
  content?: string
  type?: string
  is_default?: boolean
}

export interface CampaignTemplateListResponse {
  items: CampaignTemplate[]
  total: number
}

export interface TemplatePreviewRequest {
  contact_data: Record<string, string>
}

export interface TemplatePreviewResponse {
  subject: string
  content: string
}

// API Functions
export async function listTemplates(params?: {
  type?: string
  is_default?: boolean
}): Promise<CampaignTemplateListResponse> {
  const queryParams = new URLSearchParams()
  if (params?.type) queryParams.append('type', params.type)
  if (params?.is_default !== undefined) queryParams.append('is_default', params.is_default.toString())

  const query = queryParams.toString()
  return apiClient.get<CampaignTemplateListResponse>(`/api/master-admin/templates${query ? `?${query}` : ''}`)
}

export async function getTemplate(templateId: number): Promise<CampaignTemplate> {
  return apiClient.get<CampaignTemplate>(`/api/master-admin/templates/${templateId}`)
}

export async function createTemplate(template: CampaignTemplateCreate): Promise<CampaignTemplate> {
  return apiClient.post<CampaignTemplate>('/api/master-admin/templates', template)
}

export async function updateTemplate(templateId: number, update: CampaignTemplateUpdate): Promise<CampaignTemplate> {
  return apiClient.put<CampaignTemplate>(`/api/master-admin/templates/${templateId}`, update)
}

export async function deleteTemplate(templateId: number): Promise<void> {
  return apiClient.delete<void>(`/api/master-admin/templates/${templateId}`)
}

export async function renderTemplatePreview(templateId: number, contactData: Record<string, string>): Promise<TemplatePreviewResponse> {
  return apiClient.post<TemplatePreviewResponse>(`/api/master-admin/templates/${templateId}/preview`, {
    contact_data: contactData,
  })
}

